File: Assignments/Assign0/code/part1/src/Warmup.hs
module Warmup where

type Pos = (Int, Int)
data Direction = North | South | East | West

move :: Direction -> Pos -> Pos
move North (x,y) = (x, y+1)
move West  (x,y) = (x-1, y)
move East  (x,y) = (x+1, y)
move South (x,y) = (x, y-1)
-- move _ (x,y) = (x, y)


moves :: [Direction] -> Pos -> Pos
moves [] pos = pos
moves (d:ds) pos = moves ds . move d $ pos 


data Nat = Zero | Succ Nat
  deriving (Eq, Show, Read, Ord)

add :: Nat -> Nat -> Nat
add Zero x = x
add x Zero = x
add x (Succ y) = add (Succ x) y 

mult :: Nat -> Nat -> Nat
mult Zero _ = Zero
mult _ Zero = Zero
mult (Succ Zero) x = x
mult x (Succ Zero) = x
mult x (Succ y) = add x . mult x $ y


nat2int :: Nat -> Int
nat2int Zero = 0
nat2int (Succ x) = 1 + nat2int x

int2nat :: Int -> Nat
int2nat x
  | x==0 = Zero
  | x>0 = Succ( int2nat (x-1))
  | otherwise = Zero


data Tree = Leaf | Node Int Tree Tree
  deriving (Eq, Show, Read, Ord)

insert :: Int -> Tree -> Tree
insert n Leaf = Node n Leaf Leaf
insert n t@(Node k left right)
  | n==k = t
  | n<k = Node k (insert n left) right
  | otherwise = Node k left (insert n right)

File: Assignments/Assign4/code/part1/src/warmup.erl
-module(warmup).
-export([move/2, insert/3, lookup/2]).

% direction is one of the atoms north, south, east or west

move(north, {X, Y}) -> {X, Y+1};
move(south, {X, Y}) -> {X, Y-1};
move(east, {X,Y}) -> {X+1, Y};
move(west,  {X, Y}) -> {X-1, Y}.
% complete the definition


% A binary search tree is either
%      the atom leaf
%   or a tuple {node, Key, Value, Left, Right}
%      where Left and Right are binary search trees, and all the keys
%      in Left are smaller than Key and all the keys in Right are
%      larger than Key


% insert inserts a key and a value into a binary search tree. If the
% key is already there the value is updated.

insert(Key, Value, Tree) -> 
    case Tree of
        leaf -> {node, Key, Value, leaf, leaf};
        {node, Key, _, Left, Right} -> {node, Key, Value, Left, Right};
        {node, K, V, Left, Right} when Key<K -> {node, K, V, insert(Key, Value, Left), Right};
        {node, K, V, Left, Right} when Key>K -> {node, K, V, Left, insert(Key, Value, Right)}
    end.

% complete the definition.


% lookup find the value associated to a key in a binary search
% tree. Returns {ok, Value} if the key is in the tree; or none if the
% key is not in the tree.

lookup(Key, Tree) -> 
    case Tree of
        leaf -> none;
        {node, Key, V, _, _} -> {ok, V};
        {node, K, _, Left, _} when Key<K -> lookup(Key, Left);
        {node, K, _, _, Right} when Key>K -> lookup(Key, Right)
    end.
% complete the definition.
